### DISCUSSION TITLE
ColumnClause comparison fails with custom type when not explicitly informing type_

### DISCUSSION THREAD

**Discussion by williamjamir** [sqlalchemy/sqlalchemy#7350] [Usage Questions] [answered]
I'm opening this discussion because I don't know if I'm misusing the `column` operator or if it's really an issue.

I have a custom extension for Postgress called `ajdate` that basically "enhances" the Date type.
To use it in my codebase, I created a UserDefinedType and a class with TypeDecorator extending the Data type.

```python

class AjDateType(UserDefinedType[str]):
    def get_col_spec(self) -> str:
        return 'AJDATE'


class AjDate(TypeDecorator[date]):
    impl = AjDateType
    cache_ok = True

    def process_bind_param(self, value: Optional[date], dialect: Any) -> Optional[str]:
        if value is not None:
            return str(value)
        return None

    def process_result_value(self, value: Optional[str], dialect: Any) -> Optional[date]:
        if value is not None:
            return date.fromisoformat(value)
        return None


Base = declarative_base()

class StatsData(Base):
    date = Column(AjDate, nullable=False)

```
With the following statements, everything works fine:

```python
select(StatsData).where(StatsData.date == constants.DATE_1)
select(StatsData).where(column('date', type_=AjDate) == constants.DATE_1)
```

But when using the `column`  operator without `type_` I got the following error:
```python
>>> select(StatsData).where(column('date') == constants.DATE_1)

# psycopg2
psycopg2.errors.UndefinedFunction: operator does not exist: ajdate = date
E       LINE 3: WHERE date = '2021-01-01'::date
E                          ^
E       HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.

# asyncpg
sqlalchemy.exc.ProgrammingError: (sqlalchemy.dialects.postgresql.asyncpg.ProgrammingError) <class 'asyncpg.exceptions.UndefinedFunctionError'>: operator does not exist: ajdate = date
E                   HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.
```

My question is:
The `type_` is mandatory and expected when using custom types with the `column` operator? I was expecting that the operator would get the type defined on the class as well.

Versions
OS: Ubuntu 21.10
Python: 3.9.6
SQLAlchemy: 1.4.27
Database: Postgress 12.8

**Comment by zzzeek** [comment, marked as answer]
given:

    >>> select(StatsData).where(column('date') == constants.DATE_1)

>  I was expecting that the operator would get the type defined on the class as well.

if you name "column('date')", that object has no connection of any kind to your StatsData class, SQLAlchemy doesn't try to match up strings as this is not reliable short of implementing a full SQL parser (PostgreSQL handles that job here instead).

so given you are using a custom datatype, SQLAlhcemy is going to normally decide that a Python datetime object (I would assume that's what constants.DATE_1 is here) should be cast as a "::date" as we see here.   If your custom datatype needs the exact "ajdate" type then you need to either apply a type to `column("date", AjDate)` or use a [cast](https://docs.sqlalchemy.org/en/14/core/sqlelement.html?highlight=cast#sqlalchemy.sql.expression.cast) or [type_coerce](https://docs.sqlalchemy.org/en/14/core/sqlelement.html?highlight=cast#sqlalchemy.sql.expression.type_coerce).


much better though would be to fix the custom "ajdate" PostgreSQL extension to include an operator implementation for 
"= date" as this is a trival cast.
